### DISCUSSION TITLE
separate audio channels

### DISCUSSION THREAD

**Discussion by cxrodgers** [wehr-lab/autopilot#62] [General]
Sorry for the discussion spam today ... Is it currently supported to play a sound only from the left (or right) audio channel? I can get it to play out of both, but not out of just one.

This line suggests that this is not currently supported with jack server
https://github.com/wehr-lab/autopilot/blob/043fb6cf7c8642d4169be0823c773b358b2c4984/autopilot/stim/sound/jackclient.py#L234

Is it supported with pyo_server? I briefly attempted to switch to pyo but had some "Pyo built without Jack support" issue. Not sure if it will be easier to get jack or pyo going here. I know there's other issues with pyo anyway.

**Comment by sneakers-the-rat** [comment]
Not spam! important discussion! I was waiting for this day where we would have to implement multichannel sound ;)

Have to run to a seminar in a sec, but this actually was initially implemented, but I took it out in an act of haste because i was having indexing/shape problems and always knew it would need to be returned to.

The relevant lines are here:
On Init, should read from `OUTCHANNELS` or `NCHANNELS` (you can see I vascillated on the implementation there, where ideally one would be able to specify arbitrary output configurations rather than just number of channels, but didn't finish) to determine what to initialize. the jack terminology might be a little confusing, but basically everything is a input or output "port" so you create some virtual output 'port' that exposes an array to write to, and then connect that to the physical output 'ports' that you desire. 
https://github.com/wehr-lab/autopilot/blob/043fb6cf7c8642d4169be0823c773b358b2c4984/autopilot/stim/sound/jackclient.py#L173-L195

And then you just need to handle iterating over the multiple out ports, detecting if eg. there is a mismatch in number of outports and the dimensionality of the audio data (eg. if two channels of audio data are given and two output channels are activated, split audio data to each... vs. if one channel of audio data is given and two output channels are activated, copy it to both...)

https://github.com/wehr-lab/autopilot/blob/043fb6cf7c8642d4169be0823c773b358b2c4984/autopilot/stim/sound/jackclient.py#L314-L315

I am slammed with like ~workin on my degree requirements~ until Thursday, but I can guide until then and rejoin the coding effort then!!

**Reply by sneakers-the-rat** [reply]
whoops intended to link to the relevant docs in the Python-Jack interface: https://jackclient-python.readthedocs.io/en/0.5.3/usage.html

**Comment by cxrodgers** [comment]
Ok with these tips it's just about working now! let's talk again after Thursday

**Reply by sneakers-the-rat** [reply]
omg you already got it working? that was literally 4 minutes flat, nicely done :)
